### PR TITLE
Issue #19 and #20 - Describe auto (C++11) type inferences

### LINKED ISSUE #19: Add a new option -fdump-auto-type-inference
Add a new option `-fdump-auto-type-inference` to Clang.

### LINKED ISSUE #20: Dump auto (C++11) type inferences
C++11 provides an `auto` keyword for type inference:
https://en.cppreference.com/w/cpp/language/auto

When the option `-fdump-auto-type-inference` (as described in #19) is specified, emit an informational message that describes the type that was inferred for the `auto` keyword.

For example, with the following code:
```
auto number = 42;
auto name = std::string("John");

std::vector<int> numbers = {1, 2, 3, 4, 5};
auto iterator = numbers.begin();
```

The option should emit informational messages describing each of the types that were inferred by virtue of the `auto` keyword, something like:
```
test.c:1:1: note: type of 'number' deduced as 'int'
test.c:2:1: note: type of 'name' deduced as 'std::string'
test.c:5:1: note: type of 'iterator' deduced as 'std::vector<int>::iterator'
```

Extend the functionality for function return types and non-type template parameters as well.

### PR DESCRIPTION
**Pull request by nisarga3**

### CODE REVIEW

File: clang/include/clang/Frontend/FrontendOptions.h [hunk 1]
```diff
@@ -270,7 +270,6 @@ class FrontendInputFile {
     return *Buffer;
   }
 };
-
```
Comment (sandeepkosuri) [inline clang/include/clang/Frontend/FrontendOptions.h]: seems unnecessary

File: clang/lib/Frontend/CompilerInvocation.cpp [hunk 2]
```diff
@@ -110,7 +110,6 @@ using namespace llvm::opt;
 //===----------------------------------------------------------------------===//
 // Helpers.
 //===----------------------------------------------------------------------===//
-
```
Comment (sandeepkosuri) [inline clang/lib/Frontend/CompilerInvocation.cpp]: remove whitespace changes like this. Have you run `clang-format` on your changes?

File: clang/include/clang/Sema/Sema.h [hunk 3]
```diff
@@ -68,13 +68,17 @@
...
+#include "llvm/Support/CommandLine.h"
 #include <deque>
 #include <memory>
 #include <optional>
 #include <string>
 #include <tuple>
 #include <vector>
 
+extern llvm::cl::OptionCategory DumpAutoInference;
+extern llvm::cl::opt<bool> DumpAutoTypeInference;
```
Comment (sandeepkosuri) [inline clang/include/clang/Sema/Sema.h:83]: do these have to be global? can't they be part of `sema` class ?

File: clang/lib/Sema/SemaDecl.cpp [hunk 4]
```diff
@@ -13200,6 +13202,18 @@ bool Sema::DeduceVariableDeclarationType(VarDecl *VDecl, bool DirectInit,
...
+  SourceLocation Loc = VDecl->getLocation();
+  if (SM.isWrittenInMainFile(Loc)) {
+    if(DumpAutoTypeInference.getNumOccurrences())
+    {
+      DiagnosticsEngine &Diag = Context.getDiagnostics();
+      unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, "type of '%0' deduced as %1");
+      Diag.Report(Loc, DiagID) << VDecl->getNameAsString() << DeducedType;
+      }
+  }
+
```
Comment (sandeepkosuri) [inline clang/lib/Sema/SemaDecl.cpp]: fix indentation by running `clang-format`

File: clang/lib/Sema/SemaStmt.cpp [hunk 5]
```diff
@@ -3967,6 +3968,15 @@ bool Sema::DeduceFunctionTypeFromReturnExpr(FunctionDecl *FD,
...
 
+      // Print the deduced return type
+  SourceLocation Loc = FD->getLocation();
+  if(DumpAutoTypeInference.getNumOccurrences())
+  {
+    DiagnosticsEngine &Diag = Context.getDiagnostics();
+    unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, "function return type of '%0' deduced as %1");
+    Diag.Report(Loc, DiagID) << FD->getNameAsString() << Deduced;
+  }
+
```
Comment (sandeepkosuri) [inline clang/lib/Sema/SemaStmt.cpp]: `clang-format`

File: clang/lib/Sema/SemaStmt.cpp [hunk 6]
```diff
@@ -3967,6 +3968,15 @@ bool Sema::DeduceFunctionTypeFromReturnExpr(FunctionDecl *FD,
     // Update all declarations of the function to have the deduced return type.
     Context.adjustDeducedFunctionResultType(FD, Deduced);
 
+      // Print the deduced return type
+  SourceLocation Loc = FD->getLocation();
+  if(DumpAutoTypeInference.getNumOccurrences())
```
Comment (sandeepkosuri) [inline clang/lib/Sema/SemaStmt.cpp]: is the check
```
SM.isWrittenInMainFile(Loc)
```
 not required for functions?

File: clang/lib/Sema/SemaTemplateDeduction.cpp [hunk 7]
```diff
@@ -2020,6 +2020,27 @@ static TemplateDeductionResult DeduceTemplateArgumentsByTypeMatch(
...
+        //                     break;
+        //                 }
+        //             }
+        //         }
+        //     }
+        //     if (DeducedArgType) {
+        //         llvm::outs()<< "Deduced type for Non-Type Template Parameter: "
+        //                   << DeducedArgType->getType().getAsString() << "\n";
+        //     }
+        // }
```
Comment (sandeepkosuri) [inline clang/lib/Sema/SemaTemplateDeduction.cpp]: seems to be unnecessary artefacts

File: clang/tools/clang-installapi/Options.h [hunk 8]
```diff
@@ -46,6 +46,7 @@ struct DriverOptions {
 
   /// \brief Print verbose output.
   bool Verbose = false;
+
```
Comment (sandeepkosuri) [inline clang/tools/clang-installapi/Options.h]: unnecessary change

File: clang/lib/Sema/SemaDecl.cpp [hunk 9]
```diff
@@ -13200,6 +13202,18 @@ bool Sema::DeduceVariableDeclarationType(VarDecl *VDecl, bool DirectInit,
   VDecl->setType(DeducedType);
   assert(VDecl->isLinkageValid());
 
+    // Check if the variable declaration is in the compiled file
```
Comment (sandeepkosuri) [inline clang/lib/Sema/SemaDecl.cpp]: wdym by compiled file?

**Comment by sandeepkosuri** [comment]
Please add more helpful comments

**Comment by sandeepkosuri** [comment]
Add appropriate LIT tests, to test your changes.

File: clang/lib/Sema/SemaStmt.cpp (hunk 6 above)
Comment (nisarga3) [reply clang/lib/Sema/SemaStmt.cpp]: Added the check

File: clang/include/clang/Sema/Sema.h (hunk 3 above)
Comment (nisarga3) [reply clang/include/clang/Sema/Sema.h:83]: > do these have to be global? can't they be part of `sema` class ?

No, attempting to do so results in numerous errors. The function responsible for deducing the variable does not have a `sema` object as a parameter, making it challenging to access this option. Nevertheless, I have observed that similarly declared options are placed within the `opts` namespace. Therefore, I have made the necessary adjustments accordingly.

Comment (sandeepkosuri) [reply clang/include/clang/Sema/Sema.h:83]: Alright

**Review by sandeepkosuri — DISMISSED** [review]
LGTM

File: clang/include/clang/Driver/Options.td [hunk 10]
```diff
@@ -7474,6 +7474,12 @@ def stats_file : Joined<["-"], "stats-file=">,
 def stats_file_append : Flag<["-"], "stats-file-append">,
   HelpText<"If stats should be appended to stats-file instead of overwriting it">,
   MarshallingInfoFlag<FrontendOpts<"AppendStats">>;
+
+def fdump_auto_type_inference : Flag<["-"], "fdump-auto-type-inference">, Group<f_Group>,
+    HelpText<"Enable dumping of auto type inference information, Use -mllvm -fdump-auto-type-inference to enable.">;
```
Comment (soumitrachatterjee) [inline clang/include/clang/Driver/Options.td]:
```suggestion
    HelpText<"Dump auto type inference information">;
```

File: clang/include/clang/Driver/Options.td [hunk 11]
```diff
@@ -7474,6 +7474,12 @@ def stats_file : Joined<["-"], "stats-file=">,
...
+def fno_dump_auto_type_inference : Flag<["-"], "fno-dump-auto-type-inference">,Group<f_Group>,
+    HelpText<"Disable dumping of auto type inference information">;
```
Comment (soumitrachatterjee) [inline clang/include/clang/Driver/Options.td]:
```suggestion
    HelpText<"Disable dumping auto type inference information">;
```

Comment (soumitrachatterjee) [inline clang/include/clang/Driver/Options.td]: Make sure both these options are restricted to C++ compilation only. 
Additionally, if the C++ standard being compiled is less than C++11 (refer options `-std=c++11`, etc.), then this options should be invalid.

File: clang/lib/Sema/Sema.cpp [hunk 12]
```diff
@@ -55,6 +55,14 @@
 using namespace clang;
 using namespace sema;
 
+namespace opts {
+llvm::cl::OptionCategory DumpAutoInference("DumpAutoInference");
+llvm::cl::opt<bool> DumpAutoTypeInference{
+    "fdump-auto-type-inference",
+    llvm::cl::desc("Dump deduced type of auto type"), llvm::cl::ZeroOrMore,
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/Sema.cpp]:
```suggestion
    llvm::cl::desc("Dump compiler-deduced type for variables and return expressions declared using C++ 'auto' keyword"), llvm::cl::ZeroOrMore,
```

File: clang/lib/Sema/SemaDecl.cpp [hunk 13]
```diff
@@ -13200,6 +13202,17 @@ bool Sema::DeduceVariableDeclarationType(VarDecl *VDecl, bool DirectInit,
   VDecl->setType(DeducedType);
   assert(VDecl->isLinkageValid());
 
+  // Check if the variable declaration is in the source file
+  SourceManager &SM = getSourceManager();
+  SourceLocation Loc = VDecl->getLocation();
+  if (SM.isWrittenInMainFile(Loc) &&
+      opts::DumpAutoTypeInference.getNumOccurrences()) {
+    // Emit a remark for deduced auto variable type when option fdump-auto-type-inference is enabled
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/SemaDecl.cpp]:
```suggestion
    // Emit a remark indicating the compiler-deduced type for variables declared using the C++ 'auto' keyword
```

File: clang/lib/Sema/SemaStmt.cpp [hunk 14]
```diff
@@ -3967,6 +3968,16 @@ bool Sema::DeduceFunctionTypeFromReturnExpr(FunctionDecl *FD,
     // Update all declarations of the function to have the deduced return type.
     Context.adjustDeducedFunctionResultType(FD, Deduced);
 
+  // Emit a remark for deduced return type if the option fdump-auto-type-inference is enabled
+  SourceManager &SM = getSourceManager();
+  SourceLocation Loc = FD->getLocation();
+  if (SM.isWrittenInMainFile(Loc) &&
+      opts::DumpAutoTypeInference.getNumOccurrences()) {
+    DiagnosticsEngine &Diag = Context.getDiagnostics();
+    unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, "function return type of '%0' deduced as %1");
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/SemaStmt.cpp]:
```suggestion
    unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, "return type of function '%0' deduced as %1");
```

**Review by soumitrachatterjee — DISMISSED** [review]
Please test these changes against more complex test cases, and specifically against iterators and template types. 

Also use these changes (with the option enabled) to compile some large, modern C++ applications such as [googletest](https://github.com/google/googletest), [opencv](https://github.com/opencv/opencv), [cmake](https://gitlab.kitware.com/cmake/cmake), [eigen](https://gitlab.com/libeigen/eigen), etc.

**Comment by nisarga3** [comment]
[test.txt](https://github.com/soumitrachatterjee/llvm-project/files/15217590/test.txt)
The file test.txt has cpp code which has complex testcases for the option -fdump-auto-type-inference.

**To Run**
`./build/bin/clang++ -mllvm -fdump-auto-type-inference <filename.cpp>`

File: clang/test/Sema/fdump-auto-type-inference.cpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
+// RUN: %clang_cc1 -std=c++14 -mllvm -fdump-auto-type-inference %s
```
Comment (soumitrachatterjee) [inline clang/test/Sema/fdump-auto-type-inference.cpp:1]: This test should use c++11, since `auto` was introduced as part of C++11 and not C++14 standards.

**Review by soumitrachatterjee — COMMENTED** [review]
Additional cases/combinations to try and add (please fill in the missing types with help from @sandeepkosuri or @RitanyaB):
```
auto x = 1; //x : int
float* p;
auto x = p; //x : float*
auto* y = p; //y : float*
double f();
auto x = f(); //x : double
const auto& y = f(); //y : const double&
class R; 
R* h();
auto* x = h(); //x : R*
auto y = h(); //y : R*
int& g();
auto x = g(); //x : int
const auto& y = g(); //y : const int&
 
for (auto i = vec.begin(); i < vec.end(); i++) {
auto a = new auto(1);
//...
}

int f();
int& g();
auto y = g();    //y is of type int
auto& z = g();   //z is of type int&

auto (*p)() -> int; // declares p as pointer to function returning int
auto (*q)() -> auto = p; // declares q as pointer to function returning T
                         // where T is deduced from the type of p

for (auto const &outer_iter : StudentGrades) 
{
  std::wcout << outer_iter.first << std::endl;

  for (auto const &inner_iter : outer_iter.second)
  {
	 std::wcout << inner_iter.first << " : " << inner_iter.second << std::endl;
  }
}
```

Here's another larger example:
```
#include <iostream>
#include <utility>
 
template<class T, class U>
auto add(T t, U u) { return t + u; } // the return type is the type of operator+(T, U)
 
// perfect forwarding of a function call must use decltype(auto)
// in case the function it calls returns by reference
template<class F, class... Args>
decltype(auto) PerfectForward(F fun, Args&&... args) 
{ 
    return fun([std::forward](http://en.cppreference.com/w/cpp/utility/forward)<Args>(args)...); 
}
 
template<auto n> // C++17 auto parameter declaration
auto f() -> [std::pair](http://en.cppreference.com/w/cpp/utility/pair)<decltype(n), decltype(n)> // auto can't deduce from brace-init-list
{
    return {n, n};
}
 
int main()
{
    auto a = 1 + 2;          // type of a is int
    auto b = add(1, 1.2);    // type of b is double
    static_assert([std::is_same_v](http://en.cppreference.com/w/cpp/types/is_same)<decltype(a), int>);
    static_assert([std::is_same_v](http://en.cppreference.com/w/cpp/types/is_same)<decltype(b), double>);
 
    auto c0 = a;             // type of c0 is int, holding a copy of a
    decltype(auto) c1 = a;   // type of c1 is int, holding a copy of a
    decltype(auto) c2 = (a); // type of c2 is int&, an alias of a
    [std::cout](http://en.cppreference.com/w/cpp/io/cout) << "before modification through c2, a = " << a << '\n';
    ++c2;
    [std::cout](http://en.cppreference.com/w/cpp/io/cout) << " after modification through c2, a = " << a << '\n';
 
    auto [v, w] = f<0>(); //structured binding declaration
 
    auto d = {1, 2}; // OK: type of d is std::initializer_list<int>
    auto n = {5};    // OK: type of n is std::initializer_list<int>
//  auto e{1, 2};    // Error as of DR n3922, std::initializer_list<int> before
    auto m{5};       // OK: type of m is int as of DR n3922, initializer_list<int> before
//  decltype(auto) z = { 1, 2 } // Error: {1, 2} is not an expression
 
    // auto is commonly used for unnamed types such as the types of lambda expressions
    auto lambda = [](int x) { return x + 3; };
 
//  auto int x; // valid C++98, error as of C++11
//  auto x;     // valid C, error in C++
 
    [](...){}(c0, c1, v, w, d, n, m, lambda); // suppresses "unused variable" warnings
}
```

File: clang/test/Sema/fdump-auto-type-inference.cpp (hunk 15 above)
Comment (nisarga3) [reply clang/test/Sema/fdump-auto-type-inference.cpp:1]: 'auto' is allowed in lambda parameter only since c++14

File: clang/lib/Sema/SemaStmt.cpp [hunk 16]
```diff
@@ -3967,6 +3968,16 @@ bool Sema::DeduceFunctionTypeFromReturnExpr(FunctionDecl *FD,
...
+  // Emit a remark for deduced return type if the option fdump-auto-type-inference is enabled
+  SourceManager &SM = getSourceManager();
+  SourceLocation Loc = FD->getLocation();
+  if (SM.isWrittenInMainFile(Loc) &&
+      opts::DumpAutoTypeInference.getNumOccurrences()) {
+    DiagnosticsEngine &Diag = Context.getDiagnostics();
+    unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, "return type of function '%0' deduced as %1");
+    Diag.Report(Loc, DiagID) << FD->getNameAsString() << Deduced;
+  }
+
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/SemaStmt.cpp]: Why is this code duplicated? Can you move this into a single source and call that function appropriately?

**Comment by soumitrachatterjee** [comment]
> Additional cases/combinations to try and add (please fill in the missing types with help from @sandeepkosuri or @RitanyaB):
> 
> ```
> auto x = 1; //x : int
> float* p;
> auto x = p; //x : float*
> auto* y = p; //y : float*
> double f();
> auto x = f(); //x : double
> const auto& y = f(); //y : const double&
> class R; 
> R* h();
> auto* x = h(); //x : R*
> auto y = h(); //y : R*
> int& g();
> auto x = g(); //x : int
> const auto& y = g(); //y : const int&
>  
> for (auto i = vec.begin(); i < vec.end(); i++) {
> auto a = new auto(1);
> //...
> }
> 
> int f();
> int& g();
> auto y = g();    //y is of type int
> auto& z = g();   //z is of type int&
> 
> auto (*p)() -> int; // declares p as pointer to function returning int
> auto (*q)() -> auto = p; // declares q as pointer to function returning T
>                          // where T is deduced from the type of p
> 
> for (auto const &outer_iter : StudentGrades) 
> {
>   std::wcout << outer_iter.first << std::endl;
> 
>   for (auto const &inner_iter : outer_iter.second)
>   {
> 	 std::wcout << inner_iter.first << " : " << inner_iter.second << std::endl;
>   }
> }
> ```
> 
> Here's another larger example:
> 
> ```
> #include <iostream>
> #include <utility>
>  
> template<class T, class U>
> auto add(T t, U u) { return t + u; } // the return type is the type of operator+(T, U)
>  
> // perfect forwarding of a function call must use decltype(auto)
> // in case the function it calls returns by reference
> template<class F, class... Args>
> decltype(auto) PerfectForward(F fun, Args&&... args) 
> { 
>     return fun([std::forward](http://en.cppreference.com/w/cpp/utility/forward)<Args>(args)...); 
> }
>  
> template<auto n> // C++17 auto parameter declaration
> auto f() -> [std::pair](http://en.cppreference.com/w/cpp/utility/pair)<decltype(n), decltype(n)> // auto can't deduce from brace-init-list
> {
>     return {n, n};
> }
>  
> int main()
> {
>     auto a = 1 + 2;          // type of a is int
>     auto b = add(1, 1.2);    // type of b is double
>     static_assert([std::is_same_v](http://en.cppreference.com/w/cpp/types/is_same)<decltype(a), int>);
>     static_assert([std::is_same_v](http://en.cppreference.com/w/cpp/types/is_same)<decltype(b), double>);
>  
>     auto c0 = a;             // type of c0 is int, holding a copy of a
>     decltype(auto) c1 = a;   // type of c1 is int, holding a copy of a
>     decltype(auto) c2 = (a); // type of c2 is int&, an alias of a
>     [std::cout](http://en.cppreference.com/w/cpp/io/cout) << "before modification through c2, a = " << a << '\n';
>     ++c2;
>     [std::cout](http://en.cppreference.com/w/cpp/io/cout) << " after modification through c2, a = " << a << '\n';
>  
>     auto [v, w] = f<0>(); //structured binding declaration
>  
>     auto d = {1, 2}; // OK: type of d is std::initializer_list<int>
>     auto n = {5};    // OK: type of n is std::initializer_list<int>
> //  auto e{1, 2};    // Error as of DR n3922, std::initializer_list<int> before
>     auto m{5};       // OK: type of m is int as of DR n3922, initializer_list<int> before
> //  decltype(auto) z = { 1, 2 } // Error: {1, 2} is not an expression
>  
>     // auto is commonly used for unnamed types such as the types of lambda expressions
>     auto lambda = [](int x) { return x + 3; };
>  
> //  auto int x; // valid C++98, error as of C++11
> //  auto x;     // valid C, error in C++
>  
>     [](...){}(c0, c1, v, w, d, n, m, lambda); // suppresses "unused variable" warnings
> }
> ```

Did you get a chance to look into these?

File: clang/include/clang/Sema/Sema.h [hunk 17]
```diff
@@ -239,6 +240,11 @@ class PossiblyUnreachableDiag;
 class RISCVIntrinsicManager;
 class SemaPPCallbacks;
 class TemplateDeductionInfo;
+
+// // void DumpAutoTypeInference(SourceManager &SM, SourceLocation Loc, DiagnosticsEngine &Diag, unsigned DiagID, llvm::StringRef Name, QualType DeducedType);
```
Comment (soumitrachatterjee) [inline clang/include/clang/Sema/Sema.h]: Don't keep code commented. Delete unwanted code.

File: clang/include/clang/Sema/Sema.h [hunk 18]
```diff
@@ -239,6 +240,11 @@ class PossiblyUnreachableDiag;
...
+// template <size_t N>
+// void DumpAutoTypeInference(SourceManager &SM, SourceLocation Loc, const char (&RemarkMessage)[N], ASTContext &Context,llvm::StringRef Name, QualType DeducedType);
+
```
Comment (soumitrachatterjee) [inline clang/include/clang/Sema/Sema.h]: Delete unwanted code and lines

File: clang/include/clang/Sema/Sema.h [hunk 19]
```diff
@@ -715,6 +721,9 @@ class Sema final {
   /// Warn that the stack is nearly exhausted.
   void warnStackExhausted(SourceLocation Loc);
 
+  // void DumpAutoTypeInference(SourceManager &SM, SourceLocation Loc, DiagnosticsEngine &Diag, unsigned DiagID, llvm::StringRef Name, QualType DeducedType);
```
Comment (soumitrachatterjee) [inline clang/include/clang/Sema/Sema.h]: Delete

File: clang/lib/Sema/Sema.cpp [hunk 20]
```diff
@@ -516,6 +516,23 @@ void Sema::warnStackExhausted(SourceLocation Loc) {
   }
 }
 
+  // Emit a remark indicating the compiler-deduced type for variables declared using the C++ 'auto' keyword
+  void Sema::DumpAutoTypeInference(SourceManager &SM, SourceLocation Loc, bool isVar, ASTContext &Context, llvm::StringRef Name, QualType DeducedType) {
+      const char VarString[27] = "type of '%0' deduced as %1";
+      const char FunString[43] = "return type of function '%0' deduced as %1";
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/Sema.cpp]:
```suggestion
      const char VarString[] = "type of '%0' deduced as %1";
      const char FunString[] = "return type of function '%0' deduced as %1";
```

File: clang/lib/Sema/Sema.cpp [hunk 21]
```diff
@@ -516,6 +516,23 @@ void Sema::warnStackExhausted(SourceLocation Loc) {
...
+      if (SM.isWrittenInMainFile(Loc) && opts::DumpAutoTypeInference.getNumOccurrences()) {
+          DiagnosticsEngine &Diag = Context.getDiagnostics();
+          unsigned DiagID;
+          if (isVar) {
+            DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, VarString);
+          }
+          else {
+            DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, FunString);
+          }
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/Sema.cpp]:
```suggestion
          unsigned DiagID = Diag.getCustomDiagID(DiagnosticsEngine::Remark, isVar? VarString : FunString);
```

**Review by soumitrachatterjee — CHANGES_REQUESTED** [review]

File: clang/lib/Sema/Sema.cpp (hunk 21 above)
Comment (nisarga3) [reply clang/lib/Sema/Sema.cpp]: An error occurs because the ternary operator returns a `const char*` when employed with string literals. However, getCustomDiagID expects a reference to a constant character array `(const char (&)[N])`. Therefore, I made alternative changes in the subsequent commit.

File: clang/lib/Sema/Sema.cpp [hunk 22]
```diff
@@ -516,22 +516,23 @@ void Sema::warnStackExhausted(SourceLocation Loc) {
...
+                                 llvm::StringRef Name, QualType DeducedType) {
+  if (SM.isWrittenInMainFile(Loc) &&
+      opts::DumpAutoTypeInference.getNumOccurrences()) {
+    DiagnosticsEngine &Diag = Context.getDiagnostics();
+    unsigned DiagID = isVar ? Diag.getCustomDiagID(DiagnosticsEngine::Remark,
+                                                   "type of '%0' deduced as %1")
+                            : Diag.getCustomDiagID(
+                                  DiagnosticsEngine::Remark,
+                                  "return type of function '%0' deduced as %1");
+    ;
```
Comment (soumitrachatterjee) [inline clang/lib/Sema/Sema.cpp]: Spurious?

**Review by soumitrachatterjee — DISMISSED** [review]
Please do fix the minor issue.

File: clang/lib/Sema/Sema.cpp (hunk 22 above)
Comment (nisarga3) [reply clang/lib/Sema/Sema.cpp]: Done, Thanks

**Review by soumitrachatterjee — APPROVED** [review]
